{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1: Controls."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First I will take a look at both files trying to understand the nature of both of them and how they are organized. I want to read both files but I don't know the length of any of them so Im going to avoid the command *read* until I decifer the length or size of the files. I would do the same for both files but I asume they contain similar information. As I can see, information for one AGI code is separated by \\t (I can also see this in the type of file) and every new AGI code is separated from the last using a \\n. This way I can also see what information is carried in each colum (integer, float, varchar...). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "germplasm = open(\"Germplasm.tsv\", \"r\")\n",
    "print(germplasm.readlines()) #this will give me a list\n",
    "germplasm.seek(0) #set the pointer back to the beginning of the file\n",
    "for line in germplasm.readlines():\n",
    "    print(\"\", line)\n",
    "\n",
    "print(\"\") #print two blank lines to read information more clearly\n",
    "print(\"\")\n",
    "\n",
    "locus = open(\"LocusGene.tsv\", \"r\")\n",
    "print(locus.readlines()) \n",
    "locus.seek(0)\n",
    "for line in locus.readlines():\n",
    "    print(\"\", line)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To actually see if all Loci are following the same order, or if they are disposed in the same sequence I will compare the first colum of both files looking for the AGI code. I will *import re* to use regular expressions in order to find the AGI code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "\n",
    "germplasm.seek(0)\n",
    "for line in germplasm.readlines():\n",
    "    matchObj = re.search( r'AT\\d[G]\\d+',line) \n",
    "    #print(matchObj)\n",
    "    if matchObj:\n",
    "        AGIgerm = matchObj.group()\n",
    "        print(str(AGIgerm))\n",
    "        \n",
    "    else:\n",
    "        print (\"No match!!\")\n",
    "        \n",
    "print(\"\")\n",
    "print(\"\")\n",
    "\n",
    "\n",
    "locus.seek(0)   \n",
    "for line in locus.readlines():\n",
    "    matchObj = re.search( r'AT\\d[G]\\d+',line) \n",
    "    #print(matchObj)\n",
    "    if matchObj:\n",
    "        AGIlocus = matchObj.group()\n",
    "        #print(AGIlocus)\n",
    "        print(str(AGIlocus))\n",
    "       \n",
    "    else:\n",
    "        print (\"No match!!\")\n",
    "\n",
    "\n",
    "    \n",
    "#I close both files as I will be using them later and re-opening them\n",
    "germplasm.close()\n",
    "locus.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I am aware that this is not the best way to go about this problem, but it was the only logical thing my brain could think of. I have succesfully generated two sets of data containing the AGI codes in both files. Nonetheless whenever I try to print either one outside of the loop I only get the last term. \n",
    "\n",
    "I decided to try and insert my data into a list, but as I am not being succesfull with this either I will do it manually in order to run the last two lines of the code: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AGIgermlist = [\"AT1G01040\", \"AT1G01060\", \"AT1G01140\", \"AT1G01220\", \"AT2G03720\", \"AT2G03800\", \"AT2G04240\", \"AT2G05210\", \"AT3G02130\", \"AT3G02140\", \"AT3G02230\", \"AT3G02260\", \"AT3G02310\", \"AT3G02680\", \"AT3G02850\", \"AT3G02870\", \"AT3G03260\", \"AT4G14790\", \"AT4G15210\", \"AT4G15560\", \"AT4G15570\", \"AT4G15802\", \"AT4G15880\", \"AT4G16420\", \"AT4G16480\", \"AT5G10480\", \"AT5G10510\", \"AT5G11110\", \"AT5G11260\", \"AT5G11510\", \"AT5G12200\", \"AT5G13290\"]\n",
    "print (AGIgermlist)\n",
    "\n",
    "AGIlocuslist = [\"AT1G01040\", \"AT1G01060\", \"AT1G01140\", \"AT1G01220\", \"AT2G03720\", \"AT2G03800\", \"AT2G04240\", \"AT2G05210\", \"AT3G02130\", \"AT3G02140\", \"AT3G02230\", \"AT3G02260\", \"AT3G02310\", \"AT3G02680\", \"AT3G02850\", \"AT3G02870\", \"AT3G03260\", \"AT4G14790\", \"AT4G15210\", \"AT4G15560\", \"AT4G15570\", \"AT4G15802\", \"AT4G15880\", \"AT4G16420\", \"AT4G16480\", \"AT5G10480\", \"AT5G10510\", \"AT5G11110\", \"AT5G11260\", \"AT5G11510\", \"AT5G12200\", \"AT5G13290\"]\n",
    "print(AGIlocuslist)\n",
    "\n",
    "if AGIlocuslist == AGIgermlist:\n",
    "    print(\"True\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After manually introducing the information for both I could run the last two lines of command and I found out that both lists are the same, and therefore both files follow the same sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2: Design and create the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "8 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "8 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Exam2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LabInfo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>testing123</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('Exam2',),\n",
       " ('LabInfo',),\n",
       " ('germplasm',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',),\n",
       " ('testing123',)]"
      ]
     },
     "execution_count": 271,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database Exam2;\n",
    "%sql show databases;\n",
    "%sql use Exam2\n",
    "# %sql drop database Exam2\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_Exam2</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As expected, there are no tables in this database. So we will now create the two tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_Exam2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('Germplasm',), ('Locus',)]"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE Germplasm(Locus VARCHAR(10) NOT NULL PRIMARY KEY, Germplasm VARCHAR(200) NOT NULL, Phenotype VARCHAR(1000) NOT NULL, PubMed INTEGER NOT NULL);\n",
    "# %sql DESCRIBE Germplasm;\n",
    "# %sql drop table Germplasm\n",
    "# %sql CREATE TABLE Locus(Locus VARCHAR(10) NOT NULL PRIMARY KEY, Gene VARCHAR(10) NOT NULL, ProtLeng INTEGER NOT NULL);\n",
    "# %sql DESCRIBE Locus\n",
    "%sql show tables\n",
    "# %sql drop table Locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create the relation 1:1 I will use the *SELECT* command in order to display everything for both tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>Germplasm</th>\n",
       "        <th>Phenotype</th>\n",
       "        <th>PubMed</th>\n",
       "        <th>Locus_1</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProtLeng</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM Germplasm, Locus WHERE \\\n",
    "    Locus.Locus = Germplasm.Locus;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv #to read the tsv file\n",
    "\n",
    "#I open the files again using the DictReader function as I closed them after issuing the last commands\n",
    "file_locus = open('LocusGene.tsv',\"r\")\n",
    "tsv_locus = csv.DictReader(file_locus, delimiter=\"\\t\", quotechar='\"')\n",
    "for line in tsv_locus:\n",
    "    print(line)\n",
    "\n",
    "\n",
    "# Now I have a set of dictionaries where the Key is the name of the colum and the Values are the entry for each colum. \n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam2', #I make sure I change to the database I am working in\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit(True)\n",
    " \n",
    "for line in tsv_locus:\n",
    "    with connection.cursor() as cursor:\n",
    "        try:\n",
    "            sql = \"\"\"INSERT INTO Locus (Locus, Gene, ProtLeng)\n",
    "            VALUES ('{}', '{}', {});\"\"\".format(line[\"Locus\"], line[\"Gene\"], line[\"ProteinLength\"])\n",
    "            cursor.execute(sql)\n",
    "            connection.commit() \n",
    "        finally:\n",
    "            print(\"algo\")\n",
    "file_locus.close()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv #to read the tsv file\n",
    "\n",
    "file_germ = open('Germplasm.tsv',\"r\")\n",
    "tsv_germ = csv.DictReader(file_germ, delimiter=\"\\t\", quotechar='\"')\n",
    "for line in tsv_germ:\n",
    "    print(line)\n",
    "\n",
    "# file_germ.seek(0)\n",
    "# file_germ.readline()\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam2', #I make sure I change to the database I am working in\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit(True)\n",
    "for line in tsv_germ:\n",
    "    with connection.cursor() as cursor:\n",
    "        try:\n",
    "            sql = \"\"\"INSERT INTO Germplasm (Locus, Germplasm, Phenotype, PubMed)\n",
    "            VALUES ('{}', '{}', '{}', {});\"\"\".format(line[\"Locus\"], line[\"germplasm\"], line[\"phenotype\"], line[\"pubmed\"])\n",
    "            cursor.execute(sql) \n",
    "            connection.commit() #this forces to write into the database\n",
    "        finally:\n",
    "            print(\"algo\")\n",
    "file_germ.close()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * from Germplasm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM Locus;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I created both tables but the data for the Germplasm table is not being correctly entered as I only get the first line of the tsv file. This isn't happening for the Locus table although I used exactly the same code. This is putting me off!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a report that shows the full, joined, content of the two database tables (including a header line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%sql SELECT * FROM Germplasm, Locus WHERE \\\n",
    "    Locus.Locus = Germplasm.Locus;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Locus': 'AT1G01040', 'Germplasm': 'CS3828', 'Phenotype': 'Increased abundance of miRNA precursors.', 'PubMed': 17369351, 'Locus.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProtLeng': 332}, {'Locus': 'AT1G01060', 'Germplasm': 'lhy-101', 'Phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'PubMed': 16891401, 'Locus.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProtLeng': 290}, {'Locus': 'AT1G01140', 'Germplasm': 'SALK_058629', 'Phenotype': 'hypersensitive to low potassium media', 'PubMed': 17486125, 'Locus.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProtLeng': 223}, {'Locus': 'AT1G01220', 'Germplasm': 'SALK_012400C', 'Phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'PubMed': 18199744, 'Locus.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProtLeng': 190}, {'Locus': 'AT2G03720', 'Germplasm': 'SALK_042433', 'Phenotype': 'Multiple straight hairs', 'PubMed': 16367956, 'Locus.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProtLeng': 189}, {'Locus': 'AT2G03800', 'Germplasm': 'gek1-1', 'Phenotype': 'Ethanol hypersensitivity.', 'PubMed': 15215505, 'Locus.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProtLeng': 196}, {'Locus': 'AT2G04240', 'Germplasm': 'xerico', 'Phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'PubMed': 17933900, 'Locus.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProtLeng': 256}, {'Locus': 'AT2G05210', 'Germplasm': 'pot1-1', 'Phenotype': 'No visible phenotype.', 'PubMed': 17627276, 'Locus.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProtLeng': 221}, {'Locus': 'AT3G02130', 'Germplasm': 'rpk2-2', 'Phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'PubMed': 17419837, 'Locus.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProtLeng': 284}, {'Locus': 'AT3G02140', 'Germplasm': 'afp4-1', 'Phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'PubMed': 18484180, 'Locus.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProtLeng': 300}, {'Locus': 'AT3G02230', 'Germplasm': 'rgp1-1', 'Phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'PubMed': 21478444, 'Locus.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProtLeng': 301}, {'Locus': 'AT3G02260', 'Germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'Phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'PubMed': 17586653, 'Locus.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProtLeng': 279}, {'Locus': 'AT3G02310', 'Germplasm': 'sep2-1', 'Phenotype': 'Non-described subtle phenotype.', 'PubMed': 10821278, 'Locus.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProtLeng': 175}, {'Locus': 'AT3G02680', 'Germplasm': 'atnbs1-1', 'Phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'PubMed': 17672843, 'Locus.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProtLeng': 190}, {'Locus': 'AT3G02850', 'Germplasm': 'CS3816', 'Phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'PubMed': 17568770, 'Locus.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProtLeng': 234}, {'Locus': 'AT3G02870', 'Germplasm': 'vtc4-1', 'Phenotype': 'ascorbate deficient', 'PubMed': 16595667, 'Locus.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProtLeng': 311}, {'Locus': 'AT3G03260', 'Germplasm': 'hdg8-1', 'Phenotype': 'No visible phenotype.', 'PubMed': 16778018, 'Locus.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProtLeng': 194}, {'Locus': 'AT4G14790', 'Germplasm': 'pdd17', 'Phenotype': 'Defective pollen development.', 'PubMed': 19237690, 'Locus.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProtLeng': 312}, {'Locus': 'AT4G15210', 'Germplasm': 'bmy1-2', 'Phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'PubMed': 16297066, 'Locus.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProtLeng': 313}, {'Locus': 'AT4G15560', 'Germplasm': 'cla1-1', 'Phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'PubMed': 10982425, 'Locus.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProtLeng': 219}, {'Locus': 'AT4G15570', 'Germplasm': 'maa3', 'Phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'PubMed': 18772186, 'Locus.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProtLeng': 294}, {'Locus': 'AT4G15802', 'Germplasm': 'Athspb-2', 'Phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'PubMed': 20388662, 'Locus.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProtLeng': 254}, {'Locus': 'AT4G15880', 'Germplasm': 'esd4-2', 'Phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'PubMed': 17513499, 'Locus.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProtLeng': 265}, {'Locus': 'AT4G16420', 'Germplasm': 'prz1-1', 'Phenotype': 'Altered response to auxin and cytokinin', 'PubMed': 12747832, 'Locus.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProtLeng': 279}, {'Locus': 'AT4G16480', 'Germplasm': 'atint4-2', 'Phenotype': 'No visible phenotype.', 'PubMed': 16603666, 'Locus.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProtLeng': 284}, {'Locus': 'AT5G10480', 'Germplasm': 'pas2-3', 'Phenotype': 'Segregates 25% embryo lethal.', 'PubMed': 18799749, 'Locus.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProtLeng': 301}, {'Locus': 'AT5G10510', 'Germplasm': 'plt3-1', 'Phenotype': 'Short roots and shortened root meristem.', 'PubMed': 17960244, 'Locus.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProtLeng': 310}, {'Locus': 'AT5G11110', 'Germplasm': 'kns2', 'Phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'PubMed': 18779216, 'Locus.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProtLeng': 232}, {'Locus': 'AT5G11260', 'Germplasm': 'hy5-101', 'Phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'PubMed': 16891401, 'Locus.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProtLeng': 221}, {'Locus': 'AT5G11510', 'Germplasm': 'myb3r4-1', 'Phenotype': 'No visible phenotype.', 'PubMed': 17287251, 'Locus.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProtLeng': 336}, {'Locus': 'AT5G12200', 'Germplasm': 'pyd2-2', 'Phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'PubMed': 19413687, 'Locus.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProtLeng': 310}, {'Locus': 'AT5G13290', 'Germplasm': 'crn-1', 'Phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'PubMed': 12345678, 'Locus.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProtLeng': 189}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * FROM Germplasm, Locus WHERE Locus.Locus = Germplasm.Locus\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() #it will always return a list, and the individual result objects is a dictionary\n",
    "        print(results)\n",
    "           \n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()\n",
    "\n",
    "resultsfile = open(\"Results.tsv\",\"w\")\n",
    "for entry in results[1].keys():\n",
    "    resultsfile.write(str(entry))\n",
    "    resultsfile.write(\"\\t\")\n",
    "resultsfile.write(\"\\n\")\n",
    "for line in results:\n",
    "    for entry in line.values():\n",
    "        resultsfile.write(str(entry))\n",
    "        resultsfile.write(\"\\t\")\n",
    "    resultsfile.write(\"\\n\")\n",
    "    \n",
    "\n",
    "resultsfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a joined report that only includes the Genes SKOR and MAA3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Locus': 'AT3G02850', 'Germplasm': 'CS3816', 'Phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'PubMed': 17568770, 'Locus.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProtLeng': 234}, {'Locus': 'AT4G15570', 'Germplasm': 'maa3', 'Phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'PubMed': 18772186, 'Locus.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProtLeng': 294}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# %sql SELECT * FROM Germplasm, Locus WHERE Locus.Locus = Germplasm.Locus AND (Locus.Gene = \"SKOR\" OR Locus.Gene = \"MAA3\");\n",
    "\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * FROM Germplasm, Locus WHERE Locus.Locus = Germplasm.Locus AND (Locus.Gene = 'SKOR' OR Locus.Gene = 'MAA3')\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() #it will always return a list, and the individual result objects is a dictionary\n",
    "        print(results)\n",
    "       \n",
    "                          \n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()\n",
    "\n",
    "resultsfile = open(\"Results.tsv\",\"a\")\n",
    "resultsfile.write(\"\\nSecond Asignment\\n\")\n",
    "resultsfile.write(\"\\n\")\n",
    "for line in results:\n",
    "    for entry in line.values():\n",
    "        resultsfile.write(str(entry))\n",
    "        resultsfile.write(\"\\t\")\n",
    "    resultsfile.write(\"\\n\")\n",
    "    \n",
    "\n",
    "resultsfile.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * FROM Germplasm, Locus WHERE Locus.Locus = Germplasm.Locus AND (Locus.Gene = 'SKOR' OR Locus.Gene = 'MAA3')\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() #it will always return a list, and the individual result objects is a dictionary\n",
    "        print(results)\n",
    "       \n",
    "                          \n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
